doc/manpages/ndngroup.1
'\" t
.\"     Title: ndngroup
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.76.0 <http://docbook.sf.net/>
.\"      Date: 05/16/2013
.\"    Manual: \ \&
.\"    Source: \ \& 0.7.2
.\"  Language: English
.\"
.TH "NDNGROUP" "1" "05/16/2013" "\ \& 0\&.7\&.2" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ndngroup \- show and modify access control groups for access control of a NDNx content namespace\&.
.SH "SYNOPSIS"
.sp
\fBndngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-list
.sp
\fBndngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-listmembers \fIgroupFriendlyName\fR
.sp
\fBndngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-create \fIgroupFriendlyName\fR \fIgroupMember\fR\&...
.sp
\fBndngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-delete \fIgroupFriendlyName\fR
.sp
\fBndngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-add \fIgroupFriendlyName\fR \fIgroupMemberToAdd\fR\&...
.sp
\fBndngroup\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-remove \fIgroupFriendlyName\fR \fIgroupMemberToRemove\fR\&...
.SH "DESCRIPTION"
.sp
The \fBndngroup\fR utility lists or modifies access control groups published with NDNx\&. Groups may be referenced in access control lists (ACLs) under the NDNx access control profile\&.
.sp
Groups have short, human\-readable "friendly" names by which they are identified with this tool, although the internal identifier for a group is a digest of a public key\&.
.SH "OPTIONS"
.PP
\fB\-as\fR \fIpathToKeystore\fR
.RS 4
Specifies a file path to a keystore; this is used for signing any newly\-created content\&. Applicable to all operations\&. The default keystore is used if this is not specified\&.
.RE
.PP
\fB\-list\fR
.RS 4
List the groups that exist\&.
.RE
.PP
\fB\-listmembers\fR \fIgroupFriendlyName\fR
.RS 4
List the members of the group
\fIgroupFriendlyName\fR\&.
.RE
.PP
\fB\-create\fR \fIgroupFriendlyName\fR \fIgroupMember\fR\&...
.RS 4
Create the group named
\fIgroupFriendlyName\fR
with initial members
\fIgroupMember\fR\&...
.RE
.PP
\fB\-delete\fR \fIgroupFriendlyName\fR
.RS 4
Delete the group named
\fIgroupFriendlyName\fR\&.
.RE
.PP
\fB\-add\fR \fIgroupFriendlyName\fR \fIgroupMemberToAdd\fR\&...
.RS 4
Add the
\fIgroupMemberToAdd\fR\&... members to the group
\fIgroupFriendlyName\fR\&.
.RE
.PP
\fB\-remove\fR \fIgroupFriendlyName\fR \fIgroupMemberToRemove\fR\&...
.RS 4
Remove the
\fIgroupMemberToRemove\fR\&... members from the group
\fIgroupFriendlyName\fR\&.
.RE
.PP
\fB\-debug\fR \fIportno\fR
.RS 4
Allows attachment of an eclipse remote debugger to the utility at the specified port number\&.
.RE
.PP
\fB\-javaopts\fR \fIoptions\fR
.RS 4
Allows use of arbitrary java properties or options with the utility\&.
.RE
.SH "EXIT STATUS"
.PP
\fB0\fR
.RS 4
Success
.RE
.PP
\fB1\fR
.RS 4
Failure (syntax or usage error; retrieval error)
.RE
.SH "AUTHOR"
.sp
Philippe Golle


CHANGELOG
NDNx v0.1 (2013-08-23)
--------

Initial public release.

Based on CCNx version 0.7.2. 
https://github.com/ProjectCCNx/ccnx/tree/ccnx-0.7.2


javasrc/src/main/org/ndnx/ndn/utils/ndngroup.java
/*
 * A NDNx command line utility.
 *
 * Portions Copyright (C) 2013 Regents of the University of California.
 * 
 * Based on the CCNx C Library by PARC.
 * Copyright (C) 2010, 2012 Palo Alto Research Center, Inc.
 *
 * This work is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation.
 * This work is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details. You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

package org.ndnx.ndn.utils;

import java.io.File;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.SortedSet;
import java.util.logging.Level;

import org.ndnx.ndn.NDNHandle;
import org.ndnx.ndn.config.SystemConfiguration;
import org.ndnx.ndn.config.UserConfiguration;
import org.ndnx.ndn.impl.support.Log;
import org.ndnx.ndn.io.content.Link;
import org.ndnx.ndn.profiles.nameenum.EnumeratedNameList;
import org.ndnx.ndn.profiles.security.access.AccessDeniedException;
import org.ndnx.ndn.profiles.security.access.group.Group;
import org.ndnx.ndn.profiles.security.access.group.GroupAccessControlManager;
import org.ndnx.ndn.profiles.security.access.group.GroupManager;
import org.ndnx.ndn.profiles.security.access.group.MembershipListObject;
import org.ndnx.ndn.protocol.ContentName;

public class ndngroup {

	private static long TIMEOUT = 1000;
	private static ContentName userStorage = new ContentName(UserConfiguration.defaultNamespace(), "Users");
	private static ContentName groupStorage = new ContentName(UserConfiguration.defaultNamespace(), "Groups");

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// silence logging
		Log.setDefaultLevel(Level.WARNING);
		String extraUsage = "";

		if ((args == null) || (args.length == 0)) {
			usage(extraUsage);
		}

		int pos = 0;
		if (args[0].startsWith("[")) {
			extraUsage = args[0];
			pos++;
		}
		if (args[pos].equals("-h")) {
			usage(extraUsage);
		}
		if (args[pos].equals("-as")) {
			if (args.length < pos+2) usage(extraUsage);
			pos++;
			setUser(args[pos]);
			pos++;
		}

		if (args[pos].equals("-list")) {
			listGroups();
			System.exit(0);
		}
		else if (args[pos].equals("-listmembers")) {
			if (args.length < pos + 2) usage(extraUsage);
			pos++;
			String groupName = args[pos];
			listMembers(groupName);
			System.exit(0);
		}
		else if (args[pos].equals("-delete")) {
			if (args.length < pos + 2) usage(extraUsage);
			pos ++;
			String groupName = args[pos];
			deleteGroup(groupName);
			System.exit(0);
		}
		else if (args[pos].equals("-create") || args[pos].equals("-add") || args[pos].equals("-remove")) {
			if (args.length < pos + 2) usage(extraUsage);
			String command = args[pos];
			pos++;
			String groupName = args[pos];
			pos++;
			ArrayList<Link> groupMembers = new ArrayList<Link>();
			for (int i=pos; i<args.length; i++) {
				try {
					Link lk = new Link(ContentName.fromNative(args[i]));
					groupMembers.add(lk);
				}
				catch (Exception e) {
					e.printStackTrace();
					System.exit(1);
				}
			}
			if (command.equals("-create")) createGroup(groupName, groupMembers);
			else if (command.equals("-add")) addMember(groupName, groupMembers);
			else if (command.equals("-remove")) removeMember(groupName, groupMembers);
			System.exit(0);
		}
		else {
			usage(extraUsage);
		}

	}

	public static void usage(String extraUsage) {
		System.out.println("usage:");
		System.out.println("ndngroup " + extraUsage + "[-as pathToKeystore] -list");
		System.out.println("ndngroup " + extraUsage + "[-as pathToKeystore] -listmembers groupFriendlyName");
		System.out.println("ndngroup " + extraUsage + "[-as pathToKeystore] [-create | -add | -remove] groupFriendlyName (groupMember)*");
		System.out.println("ndngroup " + extraUsage + "[-as pathToKeystore] -delete groupFriendlyName");
		System.exit(1);
	}

	public static void setUser(String pathToKeystore) {
		File userDirectory = new File(pathToKeystore);
		String userConfigDir = userDirectory.getAbsolutePath();
		System.out.println("Loading keystore from: " + userConfigDir);
		UserConfiguration.setUserConfigurationDirectory(userConfigDir);
		// Assume here that the name of the file is the userName
		String userName = userDirectory.getName();
		if (userName != null) {
			System.out.println("User: " + userName);
			UserConfiguration.setUserName(userName);
		}
	}

	public static void listGroups() {
		try {
			EnumeratedNameList userDirectory = new EnumeratedNameList(groupStorage, NDNHandle.open());
			userDirectory.waitForChildren(TIMEOUT);
			Thread.sleep(TIMEOUT);

			SortedSet<ContentName> availableChildren = userDirectory.getChildren();
			if ((null == availableChildren) || (availableChildren.size() == 0)) {
				System.out.println("No group found in: " + groupStorage);
			}
			else {
				System.out.println(availableChildren.size() + " group(s) found in: " + groupStorage);
				for (ContentName child : availableChildren) {
					ContentName fullName = groupStorage.append(child);
					System.out.println(fullName);
				}
			}
		}
		catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static void listMembers(String groupName) {
		try{
			GroupAccessControlManager acm = new GroupAccessControlManager(null, groupStorage, userStorage, NDNHandle.open());
			GroupManager gm = acm.groupManager();
			Thread.sleep(TIMEOUT);
			Group g = gm.getGroup(groupName, SystemConfiguration.getDefaultTimeout());
			MembershipListObject ml = g.membershipList();
			LinkedList<Link> lll = ml.contents();
			System.out.println("The group " + groupName + " has " + lll.size() + " members:");
			for (Link l: lll) {
				System.out.println(l.targetName());
			}
		}
		catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
	}


	public static void createGroup(String groupName, ArrayList<Link> membersToAdd) {
		try {
			ContentName root = ContentName.fromNative("/");
			GroupAccessControlManager acm = new GroupAccessControlManager(root, groupStorage, userStorage, NDNHandle.open());
			GroupManager gm = acm.groupManager();
			gm.createGroup(groupName, membersToAdd, SystemConfiguration.getDefaultTimeout());
		} catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		System.out.println("Group " + groupName + " created, with members: ");
		for (Link lk: membersToAdd) {
			System.out.println(lk.targetName());
		}
	}

	public static void deleteGroup(String groupName) {
		try {
			GroupAccessControlManager acm = new GroupAccessControlManager(null, groupStorage, userStorage, NDNHandle.open());
			GroupManager gm = acm.groupManager();
			gm.deleteGroup(groupName);
		} catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		System.out.println("Deleted group " + groupName);
	}

	public static void addMember(String groupName, ArrayList<Link> membersToAdd) {
		try {
			GroupAccessControlManager acm = new GroupAccessControlManager(null, groupStorage, userStorage, NDNHandle.open());
			GroupManager gm = acm.groupManager();
			Thread.sleep(TIMEOUT);
			Group g = gm.getGroup(groupName, SystemConfiguration.getDefaultTimeout());
			g.modify(membersToAdd, null);
		}
		catch (AccessDeniedException aed) {
			System.out.println("You do not have the permission to edit the membership of Group " + groupName);
			System.exit(1);
		}
		catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		System.out.println("Added to group " + groupName + " the following members: ");
		for (Link lk: membersToAdd) {
			System.out.println(lk.targetName());
		}
	}

	public static void removeMember(String groupName, ArrayList<Link> membersToRemove) {
		try {
			GroupAccessControlManager acm = new GroupAccessControlManager(null, groupStorage, userStorage, NDNHandle.open());
			GroupManager gm = acm.groupManager();
			Thread.sleep(TIMEOUT);
			Group g = gm.getGroup(groupName, SystemConfiguration.getDefaultTimeout());
			g.modify(null, membersToRemove);
		}
		catch (AccessDeniedException aed) {
			System.out.println("You do not have the permission to edit the membership of Group " + groupName);
			System.exit(1);
		}
		catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		System.out.println("Removed from group " + groupName + " the following members: ");
		for (Link lk: membersToRemove) {
			System.out.println(lk.targetName());
		}
	}

}
